### ISSUE TITLE
Hifiberry amp+ is not an option in kodi

### ISSUE THREAD

**Issue by riesch** [Saiyato/volumio-kodi-plugin#9]
After install of the plugin, playback options are hdmi, analog or both. In de settings of the plugin I selected the dac playback option.

**Comment by Saiyato** [comment]
Hi, can you post the output of: `cat /proc/asound/cards`

**Comment by riesch** [comment]
Hi, I tried to reinstall kodi for this purpose, but cant get it active from volumio.


Verstuurd vanaf mijn iPad

**Comment by Saiyato** [comment]
You don't need to reinstall Kodi to post the results of `cat /proc/asound/cards`, installation of plugins will not affect the output. :)

As for your problem activating the plugin; I'm going to need a little more information:

1. What image version did you use?
2. What is shown in the browser?
3. Can you post the latest entries from the log using `sudo journalctl -f` and then starting the plugin?

**Comment by SCiunczyk** [comment]
Hi,

I have similar situation.

Result of cat /proc/asound/cards

`0 [ALSA           ]: bcm2835 - bcm2835 ALSA`
`
                     bcm2835 ALSA`
`1 [sndrpihifiberry]: HifiberryAmp - snd_rpi_hifiberry_amp`
`                      snd_rpi_hifiberry_amp`

On Volumio I have the HiFiBerry Amp option available,

![volumio](https://user-images.githubusercontent.com/22594408/28754956-68ba3904-7550-11e7-8bdd-578b64f473b8.jpg)

![volumioplugin](https://user-images.githubusercontent.com/22594408/28755070-dcb796d8-7552-11e7-91a0-95b680fe2283.jpg)
but on Kodi there is only HDMI, Analog or Both. No HiFiBerry Amp.

![screenshot002](https://user-images.githubusercontent.com/22594408/28754999-a97921b6-7551-11e7-9ed3-cee0c0db0ba0.png)

Do you have any idea how to fix this ?

**Comment by Saiyato** [comment]
Hey,

Thanks for the info, can you post the results of `cat /etc/asound.conf` please?
Also, you need to restart Kodi (just stop and start the plugin) for the changes to take effect.

**Comment by SCiunczyk** [comment]
Hi,

volumio@volumio:~$ cat /etc/asound.conf
```
#Override alsa.conf settings
                defaults.ctl.card 0
                defaults.pcm.card 0
#KODI
        defaults.ctl.card 1
        defaults.pcm.card 1
#ENDOFKODI

```
result is the same for plugin on and off

regards

Slawek

**Comment by Saiyato** [comment]
Hi,

Thank for posting back, you should disable the DAC in the Kodi plug-in settings. Then restart Kodi, I see my config conflicts for the Amp+

**Comment by Saiyato** [comment]
Can you tell me if that worked for you?

**Comment by SCiunczyk** [comment]
Hi,

Thanks for your suggestions. Tomorrow morning I will check and let you know.

**Comment by SCiunczyk** [comment]
Hi,

disabling the DAC in the Kodi plug-in settings enable ALSA setting in Kodi, but not HiFiBerry amp - it's the default bcm2835 (my hardware is RPi3)

![screenshot000](https://user-images.githubusercontent.com/22594408/28995428-41a4ca6c-79e9-11e7-9052-1684d6b8a48f.png)

this is not working for me.

If I install fresh LibreELEC - there is ALSA HiFiBerry amp option in Kodi. Maybe I can send some files to you to help you identify the problem ?

**Comment by SCiunczyk** [comment]
Hi,

any news ? How can I help with this problem?

**Comment by Saiyato** [comment]
Did you restart Kodi after you saved the dac-setting in the plugin?

Can you post the results of `cat /etc/asound.conf` and `aplay -l`? I can't test this scenario, because I don't own a hifiberry amp card unfortunately.

**Comment by SCiunczyk** [comment]
Yes sure I did.

here is my aplay - l result

```
aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: ALSA [bcm2835 ALSA], device 0: bcm2835 ALSA [bcm2835 ALSA]
  Subdevices: 8/8
  Subdevice #0: subdevice #0
  Subdevice #1: subdevice #1
  Subdevice #2: subdevice #2
  Subdevice #3: subdevice #3
  Subdevice #4: subdevice #4
  Subdevice #5: subdevice #5
  Subdevice #6: subdevice #6
  Subdevice #7: subdevice #7
card 0: ALSA [bcm2835 ALSA], device 1: bcm2835 ALSA [bcm2835 IEC958/HDMI]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: sndrpihifiberry [snd_rpi_hifiberry_amp], device 0: HifiBerry AMP HiFi tas5713-hifi-0 []
  Subdevices: 1/1
  Subdevice #0: subdevice #0

```
the only option to be able to select Hifiberry Amp in Kodi is to manually change the /etc/asound.conf

in default  section like bellow:
```
cat /etc/asound.conf
# Override alsa.conf settings
                defaults.ctl.card 1
                defaults.pcm.card 1
```
in that way I can select in Kodi ALSA Default (snd_rpi_hifiberry_amp Analog)
but still there is no sound in Kodi :-(

Just for test I've installed on side the fresh LibreELEC and my card is detected and working without any problem. Of course in Volumio is working as well .

FreshLibreELEC

with one line added to config.txt:
```
dtoverlay=hifiberry-amp
```
results:
```
cat /etc/asound.conf 
```
No "asound.conf" file at all
```
aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: sndrpihifiberry [snd_rpi_hifiberry_amp], device 0: HifiBerry AMP HiFi tas5713-hifi-0 []
  Subdevices: 0/1
  Subdevice #0: subdevice #0
```

I was trying to disable default sound "card 0:" in my RPi to have only one card. Then I was able to select it in Kodi but still no sound in Kodi. In Volumio - perfect.

**Comment by Saiyato** [comment]
I think LE (=LibreElec) might have the dtparam=audio=off setting in the config.txt and Volumio does not, this disables the onboard audio device(s).

Can you confirm this maybe? Also, can you try putting that in the boot config for Volumio (NB: You won't be able to use HDMI/analog out in Volumio anymore and it _might_ cause more trouble).

I can add another option (disable onboard sound), but I'm note sure what this does to HDMI.... But first things first, can you verify that LE has this setting and Volumio doesn't? And secondly, if you add the dtparam line does the sound work in Kodi? Please make sure DAC is _off_ in the Kodi plugin. As your card index needs to be/stay 0.

**Comment by SCiunczyk** [comment]
There is no dtparam=audio=off in LE

here is the LE config.txt:
```
################################################################################
#      This file is part of LibreELEC - http://www.libreelec.tv
#      Copyright (C) 2016 Team LibreELEC
#
#  LibreELEC is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 2 of the License, or
#  (at your option) any later version.
#
#  LibreELEC is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with LibreELEC.  If not, see <http://www.gnu.org/licenses/>.
################################################################################
# Bootloader configuration - config.txt
################################################################################

################################################################################
# Memory (System/GPU configuration )
################################################################################

# Default GPU memory split (do not change if you do not know what you are doing)
  gpu_mem=128

# Configure GPU memory based on SDRAM size - overrides above setting
  gpu_mem_256=112
  gpu_mem_512=160
  gpu_mem_1024=256

################################################################################
# For overclocking and various other settings, see:
# https://www.raspberrypi.org/documentation/configuration/config-txt.md
################################################################################
# Set 'force_turbo=1' to disable dynamic overclocking and enable overclocking always.
force_turbo=0

# Make display smaller to stop text spilling off the screen
#
# Note that the overscan settings only affect the splash screen and not Kodi.
#
# If you experience overscan/underscan issues the best solution is to adjust
# your TV settings ("full pixel", "1-1 pixel" etc.). Alternatively, there is a
# calibration menu in the Kodi GUI.
# disable_overscan=1

# Force HDMI even if unplugged or powered off
# hdmi_force_hotplug=1

# Doesn't sent initial active source message.
# Avoids bringing CEC (enabled TV) out of standby and channel switch when
# rebooting.
hdmi_ignore_cec_init=1

################################################################################
# License keys to enable GPU hardware decoding for various codecs
# to obtain keys visit the shop at http://www.raspberrypi.com
################################################################################

# decode_MPG2=0x00000000
# decode_WVC1=0x00000000

################################################################################
# End of default configuration
################################################################################

################################################################################
# Include distribution specific config file if it exists.
################################################################################
[all]
include distroconfig.txt

dtoverlay=hifiberry-amp
```
I'm able to select "standard" audio channels in LE:
![screenshot000](https://user-images.githubusercontent.com/22594408/29242096-b72cb98c-7f86-11e7-81ae-3d4642959b53.png)

Volumio - what dtparam lines you want to test? dtparam=audio=off ?

**Comment by Saiyato** [comment]
Yes I wanted to see if the dtparam=audio=off would make any difference on the Volumio image.

Also I've been reading into Open/LibreElec, this is not really comparable to Volumio. OE and LE are running on a bare arch(?) Linux variant, whereas Volumio runs on Jessie (way more bloated).

For example aplay -l returns the onboard card in Jessie, but not in LE. Therefore I think the ALSA configuration differs entirely. But I will not give up so easily, let's try to make this work....

Maybe if I got my hands on a hifiberry AMP for a good price I'd give it a good go, will keep checking the bargains ;)

**Comment by Saiyato** [comment]
You can also try to add dtdebug=1 to the /boot/config.txt

source: https://support.hifiberry.com/hc/en-us/articles/205311302-Configuring-OpenElec-with-device-tree-overlays

**Comment by SCiunczyk** [comment]
OKi I will try in few minutes.

BTW I'm very very motivated to have this configuration working. I'm building kind of HTPC for my yacht - RPi3 plus 2x1TB HDDs plus external power to from yacht 12V DC. Mostly I like to listen to music from HDD library or internet radio and obviously Volumio is perfect soft for that. With my yacht loudspeakers Volumio n RPi3+HiFiBerry Amp+ is more than perfect. This season it was working from the spring till now.... but I have also nice 42" TV on my yacht - it's LG SmartTV but Kodi is incomparable to the LG ugly interface. We know it well. So for me having team Volumio + Kodi is my dream configuration. Including "poor man" NAS.

If nothing help I can participate on purchase of HiFiBerry for you - but lets try first on my config.

This is my yHTPC (yacht Home Theatre PC ;-)
![obraz](https://user-images.githubusercontent.com/22594408/29242604-61dd0f54-7f90-11e7-97ef-8cf61fa7d091.png)

The top cover is still on my 3D printer...

**Comment by SCiunczyk** [comment]
OK tested.
options added in config.txt

```
dtparam=audio=off
dtdebug=1

```
so there is only one card - HiFiBerry Amp+ with number 0

```
aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: sndrpihifiberry [snd_rpi_hifiberry_amp], device 0: HifiBerry AMP HiFi tas5713-hifi-0 []
  Subdevices: 1/1
  Subdevice #0: subdevice #0

```
DAC is off in the Kodi plugin.

```
File: /etc/asound.conf
# Override alsa.conf settings
                defaults.ctl.card 0
                defaults.pcm.card 0
```
System restarted. Volumio is setup, and sound is working as usual.... and surprise....

Kodi still shows the "standard Pi sound devices" and the HiFiBerry Amp+. 

![screenshot001](https://user-images.githubusercontent.com/22594408/29242924-0b7b4e0a-7f95-11e7-96d1-0e6d7a99462c.png)

But nothing more. still no sound from HiFiBerry. HDMI output works fine.
There is no PULSE (Bluetooth output)

**Comment by balbuze** [comment]
Hi,
I have the same behaviour with iqaudio dac plus. It never appears in kodi, but works in volumio and aplay-l shows it. I tried miscellaneous config, blacklisted default onboard sound card but NADA. I saw it once with no sound and I don't understand why it appears.

**Comment by SCiunczyk** [comment]
Hi,

try this steps (with NON blacklisted default onboard sound card)
- set DAC is off in the Kodi plugin (Save)
- edit /etc/asound.conf
```
# Override alsa.conf settings
                defaults.ctl.card 1
                defaults.pcm.card 1
```
- reboot !

after that in my case HiFiBerry Amp - always appears but no sound :-(
If I only Off/On the plugin to restart Kodi (without reboot) the card not appears.

**Comment by SCiunczyk** [comment]
What is also interesting. If I set Kodi to keep audio device alive. There is not possible to play sound in Volumio (Volumio says that device is not accessible). Looks like Kodi is somehow accessing the HiFiBerry Amp+ card. Despite that, there is no sound from Kodi.

**Comment by balbuze** [comment]
I'm going to try. I have installed an other plugin (volsimpleequal an equalizer) that provide a Loopback device and then the signal to the DAC (iqaudio). In that case, kodi can use the loopback and then output to the dac. But the sound is sometime chopped or repeated... maybe a buffer size issue. This is a other story but just means that kodi can see other ouput than hdmi and default board in volumio...

**Comment by balbuze** [comment]
well, nothing better for me. Can't see the dac.

**Comment by Saiyato** [comment]
I made a other test on a rpi with a usb FDA (full digital amplifier).The output was ok in kodi asap I plug it. The configured volumio, and the usb output diseappears from kodi.What we can learn, the problem does not comes for i2SSo my guess (not proved ) is that volumio keep active the output and the kodi can't access it...Not sure but it needs more testing.
--

**Comment by SCiunczyk** [comment]
In the article about Pulse and Kodi: http://kodi.wiki/view/PulseAudio in section "6.5 Known issues" author write about something similar to our problem ... maybe this is some trace to investigate?

**Comment by SkacikPL** [comment]
I've a rpi3 with HiFiBerry DAC+ Pro and it's an exact same scenario.
Except the twist is, if you enable DAC output for kodi in plugin settings the audio output will break entirely on volumio (tested on version 2.246) and will not work again until you disable the option and reboot the machine.

TL;DR is the DAC support in this plugin is broken and if used it breaks DAC support in Volumio and as far as KODI is concerned, it doesn't care about it either way as the DAC is not detected regardless of whether the option is enabled or disabled.

**Comment by Saiyato** [comment]
I'm expecting an AMP soon, so then I'll be able to really debug this.

As for the remark above "the DAC support in this plugin is broken and if used it breaks DAC support in Volumio"; this is true. Only one application can claim the DAC/AMP I'm affraid. This is not something I can change, bypassing this requires the inclusion of other software (e.g. PulseAudio) and appropriate porting on both ends (Kodi and Volumio).
Also changing the default card index might be problematic for Volumio, this means the option 'Use DAC' will break a whole lot more than initially expected.

The one thing I can look into is why the HifiBerry AMP/DAC is not properly used in Kodi (Kodi only). Then I can investigate whether and how to properly release the soundcard so that Volumio can use it again. Maybe killing Kodi suffices, but I'm not sure.

Meaning, the course of action would be to close Kodi after use, so the soundcard is released for other applications (Volumio/SnapCast/etc/).

**Comment by Saiyato** [comment]
UPDATE:

Just tested with HifiBerry DiGi, I can use the soundcard in Kodi AND Volumio, just not at the same time. For example I can watch a movie and when I STOP (not pause) it, I can almost instantly listen to music using the Volumio UI.

So the above is not entirely true, at least not for the Digi. I have a DAC lying around, so I will test it with that. I'm expecting the same outcome, but we might have found a bug here.... to be continued.

**Comment by SkacikPL** [comment]
To me with DAC+ Pro it was impossible to do, once DAC was enabled for Kodi in plugin settings audio in volumio was unavailable until the option was disabled and KODI wasn't able to detect or use the DAC regardless of the option.

**Comment by Saiyato** [comment]
I just re-read your post, I think the DAC+ Pro might have different behavior than the 'normal' DAC.

Can you post your `/etc/mpd.conf`, `/etc/asound.conf` and `aplay -l` when just installing Volumio and the DAC+ Pro?

**Comment by SCiunczyk** [comment]
Hi,

maybe will be interesting to install Kodi with your script on pure Arch Linux or other Jessie ? Without Volumio. Just to test if the HiFiBerry Amp+ will work ?

I'm ready to made the test But I don't know how to do it. (how to run your script without Volumio).

**Comment by Saiyato** [comment]
Hey,

If you want to install Kodi on a bare distro you only need the install.sh script.

https://github.com/Saiyato/volumio-kodi-plugin/blob/master/install.sh

Run it like this: `sudo sh {path_to_script}.sh`

**Comment by SCiunczyk** [comment]
OK I will try ... and report the results.

**Comment by yanou75pro** [comment]
Hello, 
I have the same issue as you. 
i 've add  that to config..txt
dtparam=audio=off
dtdebug=1


volumio@volumio:~$ aplay -l
```
**** List of PLAYBACK Hardware Devices ****
card 0: sndrpihifiberry [snd_rpi_hifiberry_amp], device 0: HifiBerry AMP HiFi tas5713-hifi-0 []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```

volumio@volumio:~$ cat  /etc/asound.conf 
```
  # Override alsa.conf settings
		defaults.ctl.card 0
		defaults.pcm.card 0
```

my result is that I get a sound on kodi but very low (i can barely hear it when volume is 100% on kodi ) and nothing on volumio even if i turn down kodi  .

**Comment by Saiyato** [comment]
My HifiBerry AMP just came in and the first test is successful, I did not need to change anything.

The setup:

- Raspberry Pi 2B
- Hifiberry AMP
- Logitech wireless keyboard/mouse combo (to control Kodi)
powered using the barrel connector

What did I do?

1. Install Volumio
2. Configure the Hifiberry AMP
3. Test MPD playback (success)
4. Install Spotify plugin (volspotconnect1)
5. Test Spotify playback (success)
6. Install Kodi plugin (use defaults, no changes in the settings)
7. View Kodi sound settings: no Hifiberry AMP selectable (expected)
8. Configure Kodi (add NFS share for testing video playback)
9. Set 'use DAC' to 'ON' in the plugin settings
10. Stop plugin Kodi
11. Start plugin Kodi
12. Test playback in Volumio: still OK
13. Stop playback in Volumio
14. View Kodi sound settings; select Hifiberry AMP
15. Start video playback in Kodi: sound is going over the Hifiberry AMP, no problems or low volume
16. Stop video playback
17. Start playback in Volumio again: still OK
etc. etc.

Contents of asound.conf
```
volumio@volumioamp:~$ cat /etc/asound.conf 
#Override alsa.conf settings
		defaults.ctl.card 0
		defaults.pcm.card 0

#KODI
	defaults.ctl.card 1
	defaults.pcm.card 1
#ENDOFKODI
```

aplay -l
```
volumio@volumioamp:~$ aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: ALSA [bcm2835 ALSA], device 0: bcm2835 ALSA [bcm2835 ALSA]
  Subdevices: 8/8
  Subdevice #0: subdevice #0
  Subdevice #1: subdevice #1
  Subdevice #2: subdevice #2
  Subdevice #3: subdevice #3
  Subdevice #4: subdevice #4
  Subdevice #5: subdevice #5
  Subdevice #6: subdevice #6
  Subdevice #7: subdevice #7
card 0: ALSA [bcm2835 ALSA], device 1: bcm2835 ALSA [bcm2835 IEC958/HDMI]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: sndrpihifiberry [snd_rpi_hifiberry_amp], device 0: HifiBerry AMP HiFi tas5713-hifi-0 []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```

Contents of /boot/config.txt
```
volumio@volumioamp:~$ cat /boot/config.txt 
initramfs volumio.initrd
gpu_mem=16
max_usb_current=1
dtparam=audio=on
audio_pwm_mode=2
dtparam=i2c_arm=on
disable_splash=1

#### Volumio i2s setting below: do not alter ####
dtoverlay=hifiberry-amp

gpu_mem_1024=256
gpu_mem_512=144
gpu_mem_256=112
hdmi_force_hotplug=1
```

Contents of /proc/asound/cards
```
volumio@volumioamp:~$ cat /proc/asound/cards
 0 [ALSA           ]: bcm2835 - bcm2835 ALSA
                      bcm2835 ALSA
 1 [sndrpihifiberry]: HifiberryAmp - snd_rpi_hifiberry_amp
                      snd_rpi_hifiberry_amp
```

So in my setup it worked straight out of the box. The versions I have used are:

Volumio: 2.246
Spotify Connect plugin: 1.6.1.
Kodi plugin: 2.2.1

Can you test this sequence? Maybe there's a compatibility problem with another plugin, or maybe the barrel feed is required (Kodi uses alot more power).

**Comment by SCiunczyk** [comment]
Hi,

just did the test from list  step by step - it works !!! 

.... but only until I restart the Pi ... after restart there is no possibility to select HiFiBerry Amp+ in Kodi. = no sound via Amp+ only HDMI

Repetition of points 10-14 solve the problem.... until next restart ...

BTW I have small changes to previous hardware - now I'm using my old Pi2 - it was Pi3 before.
When I get my Pi3 back I will test it as well.

**Comment by SCiunczyk** [comment]
One more thing:

without Spotify plugin installed there is no HiFiBerry Amp+ in Kodi:

![screenshot000](https://user-images.githubusercontent.com/22594408/29743257-74c8d130-8a8e-11e7-9315-619595c55f4c.png)

**Comment by Saiyato** [comment]
Ok this looks promising though, will investigate :)

**Comment by Saiyato** [comment]
I'm not experiencing any problems after multiple reboots. The only thing that changes after a reboot is the volume, its back to 0, every single time. Changing the volume back to 50 is all I have to do for sound.

If this doesn't help, let me know. I can upload my image for you to test.

**Comment by SCiunczyk** [comment]
My hardware is (now):
- RPi 2B 1GRAM
- HiFiBerry Amp+ [https://www.hifiberry.com/products/ampplus/](https://www.hifiberry.com/products/ampplus/)

My software is:
- Volumio (volumio-2.246-2017-07-31-pi.img.zip)
- volumio-kodi-plugin.zip - from here

I have this same volume changes to zero as you, but this is not a big deal now.

If can upload your image I will test it quickly and let you know.

**Comment by Saiyato** [comment]
Here's the link to the image: https://mab.to/MpPqvqLLa

I've incorporated a test video in the Kodi installation (it's on the SD-card). Please note that I only have 16GB SD-cards, so you're gonna need at least that size (or know how to resize the image).

The image is rarred (zip has a maximum of 4GB) to make it smaller (around 1.2GB).

**Comment by SCiunczyk** [comment]
Hi @Saiyato, just tested. Results are identical like mine installation:

1. image burned into SD card - fresh start of Pi - no HiFiBerry Amp+ in Kodi:

![screenshot000](https://user-images.githubusercontent.com/22594408/29809711-764a17d8-8c9d-11e7-91a7-8c85089b6a33.png)

2. Plugin stop/start - HiFiBerry Amp+ available and set (sound veeeery low):

![screenshot001](https://user-images.githubusercontent.com/22594408/29809837-001cfc50-8c9e-11e7-8754-84d750f0e3bf.png)

3. restart of RPi - from Kodi - again no HiFiBerry Amp+ in Kodi:

![screenshot002](https://user-images.githubusercontent.com/22594408/29809889-263dda58-8c9e-11e7-97d1-2a67dcd73b3a.png)

4. Plugin stop/start and Volumio sound set to 90% - HiFiBerry Amp+ available and set (sound is OK):

![screenshot003](https://user-images.githubusercontent.com/22594408/29809920-40dbaad4-8c9e-11e7-8394-3e5b0b31c200.png)

Looks that Volumio sound gauge probably setup the global ALSA mixer volume level.

**Comment by Saiyato** [comment]
I will reflash the image tonight and retest.

How do you power your Pi? Via de micro USB connector or the barrel connector or something else?

**Comment by SCiunczyk** [comment]
That way:

![imag1957](https://user-images.githubusercontent.com/22594408/29817222-906eadd0-8cb7-11e7-8498-e05968c45565.jpg)

**Comment by Saiyato** [comment]
I have exactly the same setup. Only my board doesn't have pins extending the GPIO connectors. Like this: https://www.kiwi-electronics.nl/image/cache/data/products/raspberry-pi/addon-boards/amp_plus-1000x667.jpg

So I just reimaged my sd card and on first boot the hifiberry was not selectable in Kodi. Volumio however worked fine. Restarting the plugin fixed the soundcard problem. 

Subsequent reboots didn't change anything for me, Kodi and Volumio continued to work perfectly.

Maybe ALSA is not ready when Kodi starts... Not sure how and if I can check this... But if ALSA is not ready, the sound card is unavailable, also ALSA is in the name so it must be something like that.

The weird part is that I don't have your problem. Maybe the difference is in the SD card used. I use Kingston class 10's.

Doesn't mean that I won't fix it, I'm just trying to understand the root cause of the problem so I can implement a good fix.

Can you disable the plugin, reboot, play something in Volumio (just a few seconds), start the Kodi plug-in and see if that works for you? That may indicate a timing issue.

**Comment by SCiunczyk** [comment]
There is one thing probably different from your setup. On my testing environment I have no dedicated HDMI monitor connected to RPi. I'm using my ASUS AIO computer. So RPi always start without any monitor connected. Then I have to switch my AIO input to HDMI then I can jump to Kodi on RPi.

I will test with RPi connected to my TV and let you know. It will take few hours because I have to organize 12 V power supply in my saloon ;-)

**Comment by Saiyato** [comment]
I think I was too tired yesterday... I do have the pins on top:
![img_20170829_235349 medium](https://user-images.githubusercontent.com/12473516/29868831-c8708144-8d80-11e7-92ff-f98db66b5052.jpg)

Anyways, connecting the monitor should not really matter. I edited the /boot/config.txt to have HDMI behave as hot-swap. But it's a valid test nonetheless.

Can you describe the exact process of how you use Kodi?

Mine is as follows:

1. Boot the Pi (HDMI is connected and screen is turned on)
2. Kodi starts automatically through the plugin
3. When booted I can use the Hifiberry in both Volumio and Kodi

**Comment by Saiyato** [comment]
What is the output of your dmesg regarding the soundcard? Maybe that's where things go south.
Use the command `dmesg | grep snd`

**Comment by SCiunczyk** [comment]
Hi,
- I did the test and as you mentioned monitor not really matter.
- My process how I'm using Kodi:

1. flash image provided by you into SD card
2. connect power cord, HDMI cable, keyboard-mouse dongle, Ethernet cable
3. switch the power on
4. on my AIO computer select the active input as HDMI
5. after boot I have Kodi in front of my eye, but no HiFiBerry Amp+ available in Kodi
so...
6. on my AIO I switch to PC
7. run Firefox with volumio.local as usual
8. stop plugin
9. start plugin
10. switch back AIO input to HDMI
11. in Kodi I can select the HiFiBerry Amp+ - sound is OK

this is working fine - both Kodi and Volumio until next reset.

result of dmesg:

```
volumio@volumio:~$ dmesg | grep snd
[   11.423955] snd_bcm2835: unknown parameter 'index' ignored
[   11.671287] snd-hifiberry-amp soc:sound: ASoC: CPU DAI (null) not registered - will retry
[   11.708321] snd-hifiberry-amp soc:sound: ASoC: CODEC DAI tas5713-hifi not registered - will retry
[   12.960623] snd-hifiberry-amp soc:sound: tas5713-hifi <-> 3f203000.i2s mapping ok

```

this same result I have for HiFiBerry Amp+ available and not available  in Kodi.

**Comment by Saiyato** [comment]
Ok, so in short, the soundcard is not available in Kodi upon boot, this is something I need to investigate. The easiest way is to keep reflashing my image, because the first boot showed this behaviour.

Just to check, the sound is working fine in Volumio? After each boot? 

This narrows the problem down to the Kodi issue.

**Comment by Saiyato** [comment]
Ok, I might just have found a fix to force Kodi to use ALSA, can you edit your systemd unit (/etc/systemd/system/kodi.service) to this:

```
# Kodi as-a-service
[Unit]
Description = Kodi Media Center

# if you don't need the MySQL DB backend, this should be sufficient
After = systemd-user-sessions.service network.target sound.target

# if you need the MySQL DB backend, use this block instead of the previous
#After = systemd-user-sessions.service network.target sound.target mysql.service
#Wants = mysql.service

[Service]
User = kodi
#Group = root
Type = simple
#PAMName = login # you might want to try this one, did not work on all systems
Environment="AE_SINK=ALSA"
ExecStart = /usr/bin/kodi-standalone -- :0 -nolisten tcp vt7
Restart = on-abort
RestartSec = 5

[Install]
WantedBy = multi-user.target
```

Note: I added the Environment parameter (AE_SINK), this should force Kodi to use ALSA output (your default card) as opposed to have Kodi decide which card is used.

**Comment by SCiunczyk** [comment]
Volumio sound is always working fine - via HiFiBerry Amp+ card (of course).
The fix you suggest doesn't help. I have this same symptoms.

What is your plugin doing when we run stop then start operation in Volumio ? 

Maybe automatic plugin start with Volumio startup after reset is performed to fast/early - before system stabilise and ALSA subsystem is available to Kodi ? Maybe Volumio is "somehow blocking" the ALSA driver for little time after start - e.g. initial welcome sound ?

One observation: when Volumio is playing something e.g. internet radio and I stop/start plugin then stop playing in Volumio and switch to Kodi - HiFiBerry Amp+ card is not available in Kodi.

Only if Volumio playback is paused/stopped - the operation of start/stop make HiFiBerry Amp+ card is available in Kodi.

```
Volumio: ++++++++-----------+++++++----
Plugin:  on.........off/on.............
                                         Kodi ->> ALSA available 
```
but in situation:
```
Volumio: +++++++++++++++++++++++-------
Plugin:  on.........off/on.............
                                         Kodi ->> ALSA not available 
```

**Comment by Saiyato** [comment]
Ok, one last test then, if you stop the Kodi plugin and then reboot. Then Kodi should not start automatically, if the UI is visible and you start Kodi; is everything ok?

This should confirm a timing issue.

**Comment by balbuze** [comment]
have you try to set : "boot_priority":10 in package.json to force the plugin to load in very last place ?

**Comment by SCiunczyk** [comment]
If I follow scenario:

- restart system with plugin - off
- open volumio.local in browser
- start plugin
- set volume to some level (50%) - (not touching "play" in volumio)
- switch to Kodi
- TADAaaaaa - works as expected - sound goes via HiFiBerry Amp+

**Comment by SCiunczyk** [comment]
with "boot_priority":10 in package.json the situation is little better. After restart Kodi is able to see the ALSA card but not HiFiBerry Amp+. Only the standard Bluetooth output
![screenshot004](https://user-images.githubusercontent.com/22594408/29939039-e13c33e8-8e8a-11e7-88d5-cd4407ed7bb5.png)

**Comment by balbuze** [comment]
and disabling startup sound ?

**Comment by Saiyato** [comment]
Good find balbuze! :) Thanks.

I had to change some work to do and tomorrow night is also no option, but I'll try to do some tests too this weekend.

**Comment by SCiunczyk** [comment]
Hi guys,

I did final test ;-) 
- fresh reflash image provided by @Saiyato into SD Card
- disable startup sound in Volumio (only change I made)

Now its working as expected. No plugin "boot_priority" change , no AE_SINK parameter in kodi.service.

So I think you can leave it as is or find the way to "un-glue" Volumio from HiFiBerry Amp+ ALSA driver during plugin startup.

Have a nice weekend.

**Comment by Saiyato** [comment]
Really? Is the startup sound the culprit? Damn!
Makes me wonder why I don't have this problem on my setup... oh well, good to know, no I can look for workarounds/fixes.

Thanks for your effort in finding the culprit.

**Comment by SCiunczyk** [comment]
Well the effort was for myself as well. 
Now I have my YTPC of my dreams

**Comment by SCiunczyk** [comment]
And little bonus.... Polish translation. Ready to merge ;-)

**Comment by Saiyato** [comment]
I merged the Polish translations, thanks! I did some editing in the translation (minor), if you want you can send me a new file, if it's ok then you don't need to do anything ;)

**Comment by SCiunczyk** [comment]
It's OK

**Comment by Saiyato** [comment]
Awesome! I have not been presented any issues regarding the HifiBerry Amp, will try to figure out a neat way to inform the user when HifiBerry is used.

**Comment by erikfx** [comment]
Hi there!
I'm running the hifiberry dac on rpi3 and have the same issue that the soundcard is just available after disabling the kodi plugin in volumio browser settings and starting it again (Thanks for the intense bug research in here and the workaround). 
Unfortunately I can't save the setting "enabling dac for kodi" in my volumio kodi plugin. When I hit save and come back to the plugin the setting is disabled again like nothing was changed. 
Am I missing the trick to save the settings permanently?

**Comment by kiddzie** [comment]
Hello,
Bro Saiyato or anybody, please post again the image as the https://mab.to/MpPqvqLLa seems to point me to another site.
My gadget: rPI 2, hifiberry amp+

Nevermind, got it @ https://github.com/Saiyato/volumio-kodi-plugin/raw/master/volumio-kodi-plugin.zip
Tq Saiyato

**Comment by Saiyato** [comment]
The not being able to save problems are related to #28, I have created a new PR for the matter: https://github.com/volumio/volumio-plugins/pull/314
